### PR TITLE
[WIP] Add Training for Data with Folds; Do full training on test and valid before final testing

### PR DESCRIPTION
**Pull request by khumairraj**
This will

- Add methods to train and test on folds. Leave One out cross-validation strategy ( at a fold level ) will be applied. Each fold will be considered for testing, while all other folds will be used for training. This will lead to `n` sets of `data_splits`. Result for testing of each fold will be provided along with the aggregated result.
- Also, best grid point in the above case of `folds setup`, will only be found with the first `data_splits`. This will correspond to testing on the first fold(fold0), validation on second fold(fold1) and training on all other folds(fold2..). ( Sort the folds before deciding order(first or second) of the fold.
- For data with explicit `test, train and valid` splits defined,  best grid point will be found using the whole data.
- In both the above cases (data with folds or not with folds), the grid point will first be found followed by using this best grid point configuration to train on the `train+valid` set. Thereafter, the final testing scores will be computed.

### CODE REVIEW

**Comment by turian** [comment]
"
In both the above cases (data with folds or not with folds), the grid point will first be found followed by using this best grid point configuration to train on the train+valid set. Thereafter, the final testing scores will be computed."

This should be optional, otherwise we are breaking the behavior that we used for the open tasks.

File: heareval/predictions/task_predictions.py [hunk 1]
```diff
@@ -709,15 +714,40 @@ def dataloader_from_split_name(
...
-        metadata=metadata,
-    )
+    """
+    Get the dataloader for a `split_name` or a list of `split_name`
+
+    For a list of `split_name`, the dataset for each split will be concatenated.
+    The dataloader will be built from this concatenated
+    dataset
+    """
+    if isinstance(split_name, (list, set)):
```
Comment (turian) [inline heareval/predictions/task_predictions.py:724]: This is a bit weird, wouldn't it be better maybe to do this as two explicit functions?

File: heareval/predictions/task_predictions.py [hunk 2]
```diff
@@ -749,30 +779,153 @@ def dataloader_from_split_name(
...
 
 
-def task_predictions_train(
+def task_predictions_train_test_full(
+    best_grid_point: GridPointResult,
     embedding_path: Path,
     embedding_size: int,
-    grid_points: int,
     metadata: Dict[str, Any],
+    data_splits: Dict[str, List[str]],
```
Comment (turian) [inline heareval/predictions/task_predictions.py:802]: This adds a lot of code but it's not clear what's the difference between this and below. That makes it quite confusing and I'm worried about the code duplication

File: heareval/predictions/task_predictions.py [hunk 3]
```diff
@@ -1000,46 +1208,89 @@ def print_scores(grid_point_results: List[GridPointResult]):
...
-    test_results.update(
-        {
-            "validation_score": best_grid_point.validation_score,
-            "hparams": hparams_to_json(best_grid_point.hparams),
-            "postprocessing": best_grid_point.postprocessing,
-            "epoch": best_grid_point.epoch,
-            "time_in_min": best_grid_point.time_in_min,
-            "score_mode": best_grid_point.score_mode,
-            "embedding_path": str(embedding_path),
+    # Train the model using the best grid point configuration on the train+valid
```
Comment (turian) [inline heareval/predictions/task_predictions.py:1218]: This long function is getting a bit confusing, maybe we break it down into several functions? I am not sure tho. I am wondering if there is a way to do more code re-use to simplify things

**Comment by turian** [comment]
We might want an alternate mode where, for each train/val/test set, we do a full grid search. (Not one grid search and keep the grid point for the other folds.) This might be useful for small datasets.

File: heareval/predictions/task_predictions.py [hunk 4]
```diff
@@ -784,18 +937,20 @@ def task_predictions_train(
     start = time.time()
     predictor: AbstractPredictionModel
     if metadata["embedding_type"] == "event":
-        validation_target_events = json.load(
-            embedding_path.joinpath("valid.json").open()
-        )
-        test_target_events = json.load(embedding_path.joinpath("test.json").open())
+        # Only validation target event is required while searching for best grid point
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:940]: Are you sure? Don't we want to include early stopping on validation when we're training the final model as well?

File: heareval/predictions/task_predictions.py (hunk 2 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:802]: Agreed - I'm not really sure we need two different train functions. I understand the idea was to train the final model using the train + val to maximize the data we're using. But I think we still want to use the validation when training the final model and use early stopping.

File: heareval/predictions/task_predictions.py (hunk 3 above)
Comment (jorshi) [reply heareval/predictions/task_predictions.py:1218]: Yes. Can we instead treat everything as a folded dataset? And datasets that have predefined train, test, validation are just a special case where there is a single configuration? I think we could get rid of some of these conditionals that way.

**Review by jorshi — COMMENTED** [review]
This is looking cool - I generally agree with Joseph's comments regarding code duplication. I think there is a close solution that is more simple.

File: heareval/predictions/task_predictions.py [hunk 5]
```diff
@@ -749,30 +779,153 @@ def dataloader_from_split_name(
...
+    Model configurations are copied from the best grid point which was found using the
+    `train` and the `valid` split.
+    """
+    start = time.time()
+
+    # Get the configuration of the best_grid_point and use it to train on the
+    # full train+valid set and test on the test set
+    conf = best_grid_point.conf
+    # Set the max_epochs to the number of epochs done in the best grid point
+    conf["max_epochs"] = best_grid_point.epoch
```
Comment (jorshi) [inline heareval/predictions/task_predictions.py:823]: I see. Instead of using early stopping we cap the number of epochs to the number of epochs reached during the grid point search. Are we sure that's what we want? This is definitely quicker. But, I'm not positive this is what we want to do.

Comment (jorshi) [reply heareval/predictions/task_predictions.py:823]: @turian

Comment (turian) [reply heareval/predictions/task_predictions.py:823]: @jorshi @khumairraj No this is not correct.

You generally fix max_epochs when you are retraining and combining train+val so that you have a bigger train and no val any more.

If we are doing k-fold and fixing the hyperparameters, but we have separate train and validation folds, we should use our validation fold for early stopping again.

**Comment by khumairraj** [comment]
Inherited by #315 . Closing this
